core/src/main/scala/akka/kafka/internal/ConsumerStage.scala

```diff
@@ -0,0 +1,267 @@
+        }
+      }
+
+      val pollCallback = getAsyncCallback[Unit] { _ => schedulePoll() }
+
+      private def pushMsg(msg: Consumer.CommittableMessage[K, V]): Unit = {
+        logger.trace("Push element {}", msg)
+        push(out, msg)
+      }
+
+      setHandler(out, new OutHandler {
+        override def onPull(): Unit = {
+          if (!buffer.hasNext) poll()
+          else pushMsg(buffer.next())
+        }
+      })
+
+      private val decrementConfirmation = getAsyncCallback[Unit] { _ =>
+        awaitingConfirmation -= 1
+        logger.trace(s"Commits in progress {}", awaitingConfirmation.toString)
+      }
+
+      // impl of GraphStage.Committer that is called from the outside via CommittableOffsetImpl
+      override def commit(offset: Consumer.Offset): Future[Done] = {
+        // committed offset should be the next message the application will consume, i.e. +1
+        commit(Map(new TopicPartition(offset.topic, offset.partition) ->
+          new OffsetAndMetadata(offset.offset + 1)))
+      }
+
+      // impl of GraphStage.Committer that is called from the outside via CommittableOffsetBatchImpl
+      override def commit(offsets: Map[TopicPartition, OffsetAndMetadata]): Future[Done] = {
+        // not important that we complete with failure during stopping race,
+        // it will be completed with TimeoutException
+        if (stoppedPromise.isCompleted)
+          Future.failed(new IllegalStateException("ConsumerStage is stopped, offset commit not performed"))
+        else {
+          val done = Promise[Done]()
+          val c = materializer.scheduleOnce(COMMIT_TIMEOUT, new Runnable {
+            override def run(): Unit = {
+              done.tryFailure(new TimeoutException(
+                s"Offset commit timed out after ${COMMIT_TIMEOUT.toMillis} ms, $offsets"
+              ))
+            }
+          })
+          done.future.onComplete(__ => c.cancel())(materializer.executionContext)
+          commitCallback.invoke((offsets, done))
+          done.future
+        }
+      }
+
+      val commitCallback = getAsyncCallback[(Map[TopicPartition, OffsetAndMetadata], Promise[Done])] {
+        case (offsets, done) => commitInternal(offsets, done)
+      }
+
+      private def commitInternal(offsets: Map[TopicPartition, OffsetAndMetadata], done: Promise[Done]): Unit = {
+        awaitingConfirmation += 1
+        logger.trace(s"Start commit {}. Commits in progress {}", offsets, awaitingConfirmation.toString)
+        consumer.commitAsync(offsets.asJava, new OffsetCommitCallback {
+          override def onComplete(offsets: util.Map[TopicPartition, OffsetAndMetadata], exception: Exception): Unit = {
+            if (exception == null)
+              done.success(Done)
+            else
+              done.failure(exception)
+            logger.trace(s"Commit completed {} - {}", offsets, done.future.value)
+            decrementConfirmation.invoke(())
+          }
+        })
+        poll()
```

Comment:
This can be optimized to scheduling poll after https://issues.apache.org/jira/browse/KAFKA-3412 will be fixed